### PR TITLE
Provenance Tagging

### PR DESCRIPTION
**Pull request by rknop**
Human-readable tags associated with provenances.  (Issue #337 )

### CODE REVIEW

File: tests/webap/test_webap.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        # Now imagesdiv should have information about the sources
+        tabcontentdiv = browser.find_element( By.XPATH, "html/body/div/div/div/div/div/div/div[2]" )
+        imagesdiv = tabcontentdiv.find_element( By.XPATH, "./div/div/div/div[2]/div" )
+
+        sourcestab = imagesdiv.find_element( By.TAG_NAME, 'table' )
+        rows = sourcestab.find_elements( By.TAG_NAME, 'tr' )
+        assert len(rows) == 7
+        # check stuff about the rows?
+
+        # There is probably more we should be testing here.  Definitely.
```
Comment (whohensee) [inline tests/webap/test_webap.py:146]: We can robustify these tests down the line I think, but these look like a good starting point.

File: webap/seechange_webap.py [hunk 2]
```diff
@@ -84,46 +102,114 @@ def exposures():
...
+        else:
+            q += ( 'LEFT JOIN ( '
+                   '  SELECT re.exposure_id, re.success, re.error_message '
+                   '  FROM reports re '
+                   '  INNER JOIN provenance_tags rept ON rept.provenance_id=re.provenance_id AND rept.tag=%(provtag)s '
+                   ') r ON r.exposure_id=t.id '
+                  )
+            subdict['provtag'] = data['provenancetag']
+        # I wonder if making a primary key on the temp table would be more efficient than
+        #    all these columns in GROUP BY?  Investigate this.
```
Comment (whohensee) [inline webap/seechange_webap.py:208]: I feel like I could take an entire day just studying this monstrous 3-part query! sqlalchemy has shielded me from pure SQL, mostly...

File: webap/static/seechange.js [hunk 3]
```diff
@@ -162,16 +192,27 @@ seechange.ExposureList.prototype.render_page = function()
...
-    table = rkWebUtil.elemaker( "table", this.listdiv, { "classes": [ "exposurelist" ] } );
+    if ( this.provtag == null ) {
+        h2.appendChild( document.createTextNode( " including all provenances" ) );
+    } else {
+        h2.appendChild( document.createTextNode( " with provenance tag " + this.provtag ) );
+    }
+
+    rkWebUtil.elemaker( "p", this.listdiv,
+                        { "text": '"Detections" are everything found on subtratcions; ' +
+                          '"Sources" are things that passed prelminary cuts.' } )
```
Comment (whohensee) [inline webap/static/seechange.js:203]:
```suggestion
                        { "text": '"Detections" are everything found on subtractions; ' +
                          '"Sources" are things that passed preliminary cuts.' } )
```

Comment (whohensee) [reply webap/static/seechange.js:203]: it wouldn't be a PR without a typo!

**Review by whohensee — APPROVED** [review]
I think it looks good, no major issues found. It's nice to have this working, I think something feels very official about having the capability to have these version tags. Soon we will decide on the "launch" parameters...

File: tests/webap/test_webap.py (hunk 1 above)
Comment (rknop) [reply tests/webap/test_webap.py:146]: Lots of stuff is going to be added to the webap, so tests will be added then.

File: webap/seechange_webap.py (hunk 2 above)
Comment (rknop) [reply webap/seechange_webap.py:208]: Yeah, these things get kind of nasty.

Adding the provenance tags made these queries much more complicated.

**Pull request merged**